wagtail/contrib/modeladmin/views.py

```diff
@@ -178,20 +182,47 @@ def form_invalid(self, form):
 
 
 class InstanceSpecificView(WMABaseView):
-
     instance_pk = None
-    pk_quoted = None
-    instance = None
-
-    def __init__(self, model_admin, instance_pk):
-        super(InstanceSpecificView, self).__init__(model_admin)
-        self.instance_pk = unquote(instance_pk)
-        self.pk_quoted = quote(self.instance_pk)
-        filter_kwargs = {}
-        filter_kwargs[self.pk_attname] = self.instance_pk
-        object_qs = model_admin.model._default_manager.get_queryset().filter(
-            **filter_kwargs)
-        self.instance = get_object_or_404(object_qs)
+    pk_url_kwarg = 'instance_pk'
+
+    def __init__(self, *args, **kwargs):
+        super(InstanceSpecificView, self).__init__(*args, **kwargs)
+        if 'instance_pk' in kwargs:
+            warnings.warn(
+                "'instance_pk' should no longer be passed to %s's as_view() "
+                "method. It should be passed as a keyword argument to "
+                "the 'view' method returned by as_view() instead" %
+                self.__class__.__name__, category=RemovedInWagtail113Warning
+            )
+
+    @cached_property
+    def instance(self):
+        """
+        Return the result of self.get_model_instance() and cache it to avoid
+        repeat database queries
+        """
+        return self.get_model_instance()
+
+    def get_model_instance(self):
+        """
+        Returns an instance of self.model identified by URL parameters in the
+        current request. Raises Http404 if no match is found.
+        """
+        pk = self.kwargs.get(self.pk_url_kwarg, self.instance_pk)
+        queryset = self.model._default_manager.all().filter(pk=unquote(pk))
+        try:
+            # Get the single item from the filtered queryset
+            obj = queryset.get()
+        except self.model.DoesNotExist:
+            raise Http404(_("No %(verbose_name)s found matching the query") %
+                          {'verbose_name': self.verbose_name})
+        return obj
+
+    @property
+    def pk_quoted(self):
+        return quote(
+            self.kwargs.get(self.pk_url_kwarg, self.instance_pk)
+        )
```

Comment:
The pk as retrieved from `self.kwargs[self.pk_url_kwarg]` / `self.instance_pk` is the quoted version, isn't it? (Hence why it has to be un-quoted in line 212 above.) If so, we shouldn't be calling `quote(...)` on it again here.

The start of `get_model_instance` can then be rewritten to avoid duplication and make the logic a bit clearer:

        pk = unquote(self.pk_quoted)
        queryset = self.model._default_manager.all().filter(pk=pk)